infra/DockerRepoManager.py

```diff
@@ -0,0 +1,184 @@
+    repo_url = self._infer_main_repo()
+    self.src_on_image = os.path.join('/src', project_name)
+    super().__init__(repo_url)
+    if self.repo_name != project_name:
+      raise DockerRepoManagerError(
+          'Error, the project name must be the same as the ' +
+          'git repo name but are %s and %s' % (project_name, super().repo_name))
+
+  def cleanup(self):
+    """Removes old  TEMP_CONTAINER."""
+    self._run_command(['docker', 'stop', self.TEMP_CONTAINER])
+    self._run_command(['docker', 'container', 'rm', self.TEMP_CONTAINER])
+
+  def set_image_commit(self, commit):
+    """Creates a docker image with a specified commit as its source.
+
+    Args:
+      commit: The SHA the source is to be checked out at
+
+    Raises:
+      DockerRepoManagerError: when the commit is not successfully
+      mounted to the image
+    """
+    self.cleanup()
+    # Remove all previous images
+    self._run_command(['docker', 'rmi', self.docker_image])
+
+    # Build builder image
+    _build_image(self.project_name)
+    self.checkout_commit(commit)
+    mount_command = [
+        'docker', 'create', '--name', self.TEMP_CONTAINER, self.docker_image
+    ]
+    self._run_command(mount_command)
+
+    # Start the container to be modified
+    self._run_command(['docker', 'start', self.TEMP_CONTAINER])
+
+    # Remove outdated source repo from container
+    remove_command = [
+        'docker', 'exec', self.TEMP_CONTAINER, 'rm', '-rf', self.src_on_image
+    ]
+    self._run_command(remove_command)
+
+    # Copy updated source repo to container
+    copy_command = [
+        'docker', 'cp',
+        os.path.join(self.full_path, '.'),
+        self.TEMP_CONTAINER + ':' + self.src_on_image
+    ]
+    self._run_command(copy_command)
+
+    # Overwrite current image with new container mount
+    commit_command = [
+        'docker', 'commit', self.TEMP_CONTAINER, self.docker_image
+    ]
+    self._run_command(commit_command)
+
+    # Check the command executed correctly
+    image_commit = self.get_image_commit()
+    if not image_commit == commit:
+      raise DockerRepoManagerError(
+          'Docker commit checkout current: %s desired: %s' % (image_commit,
+                                                              commit))
+
+  def get_image_commit(self):
+    """Gets the current commit SHA from the docker image.
+
+    Returns:
+      Commit sha or None on error
+    """
+    get_commit_command = [
+        'docker', 'run', '--rm', '-it', self.docker_image, 'git', '--git-dir',
+        os.path.join(self.src_on_image, '.git'), 'rev-parse', 'HEAD'
+    ]
+
+    out, err = self._run_command(get_commit_command)
+    if err is not None:
+      return None
+    return str(out).rstrip()
+
+  def _get_docker_image_name(self):
+    """Gets the name of the docker build image for specified repo.
+
+    Returns:
+      The name of the docker build image
+    """
+    is_base_image = _is_base_image(self.project_name)
```

Comment:
I don't think we will ever be building the main image here. I think this code might be copied right? (frequently a bad idea)